File: 1090-largest-values-from-labels/1090-largest-values-from-labels.py
class Solution:
    def largestValsFromLabels(self, values: List[int], labels: List[int], numWanted: int, useLimit: int) -> int:
        

        zipped = list(zip(values, labels))
        

        zipped.sort(key = lambda x: -x[0])
        print(zipped)
        
        total = 0
        i = 0
        n = len(values)
        used = defaultdict(int)
        while i < n and numWanted > 0:
            x = zipped[i][0]
            label = zipped[i][1]

            if used[label] < useLimit:
                total += x
                used[label] += 1
                numWanted -= 1
                i += 1
            else:
                i += 1
        
        return total

File: 1090-largest-values-from-labels/1090-largest-values-from-labels.cpp
class Solution {
public:
    int largestValsFromLabels(vector<int>& values, vector<int>& labels, int numWanted, int useLimit) {
        int n = (int)values.size();

        // pair as (value, label)
        vector<pair<int, int>> items;
        
        for (int i = 0; i < n; i++) {
            items.push_back({values[i], labels[i]});
        }

        sort(items.begin(), items.end(), [](const auto& a, const auto& b) {
            return a.first > b.first;
        });

        unordered_map<int, int> used;
        long long sum = 0;
        int picked = 0;

        for (auto &it : items) {
            if (picked == numWanted) break;

            int val = it.first;
            int lab = it.second;

            if (used[lab] < useLimit) {
                used[lab]++;
                sum += val;
                picked++;
            }
        }
        return (int)sum;
    }
};